Split the generated tags into different traits

JPonte opened the pull request:
@davesmith00000 Not sure if this is what you had in mind in the comments in #96? 

The tags are split into different traits and all merged into the `HtmlTags` in the end.

davesmith00000 commented:
Looks promising! I've kicked off the CI build. I think I'll need to clone it to review it properly. Will do that as soon as I can.

davesmith00000 commented:
Cloned it, certainly looks like it works - good job!

Where did the svg tag list come from dare I ask?

I'm going to ponder the structure a little, I'm kinda wondering if it would be good to `import tyrian.SVG.*` for example?

In the meantime, I feel like we need a tiny demo so we can see it working. Do you feel like building something into the sandbox maybe? Update of the [clock](https://github.com/PurpleKingdomGames/tyrian/blob/main/examples/clock/src/main/scala/example/Main.scala) perhaps? The sandbox is a bit of a beast so if not interested in doing that I'd quite understand. I can always build an example later.

JPonte commented:
From here: https://developer.mozilla.org/en-US/docs/Web/SVG/Element
I didn't include any of the deprecated tags or the ones without any doc page.

Having `import tyrian.SVG.*` would be interesting although I'd have to think better how to make it work with the whole trait situation.

I'll update the clock and take a look at the sandbox. Is the sandbox a big example with all of the features in it? Didn't realise it existed before.

aumann commented:
Just as an aside: +1 for the SVG.* namespace idea. I would offer to build on your approach and add aria-* tags afterwards.

davesmith00000 commented:
@JPonte - please don't update the clock example. 😄 

You could copy the clock example into the sandbox and update it, but the examples are based on the currently published version.

The sandbox and the indigo-sandbox sub projects are there so that during development you can quickly try stuff based on the branch you're working on. It's very convenient for just trying things out and serves as a poor man's manual regression suite. To use them, do `sandboxBuild` in sbt, and then in another terminal tab cd to tyrian/sandbox and run `yarn start`.

davesmith00000 commented:
> I would offer to build on your approach and add aria-* tags afterwards.

Did not know about the aria tags! Yes that would be great @aumann. Worth raising an issue so we don't forget?

JPonte commented:
Ah gotcha. I was already tinkering the clock with the new IO stuff 😅
I'll add something to the sandbox then.

JPonte commented:
Alright, added the clock to a new page in the sandbox and split the SVG into its own object.
The svg tags can all be found in `tyrian.SVG.*`. I left all the attribute and tag util methods in `tyrian.Html.*` though, as well as the root tags. I guess it makes sense to always need to import that one.

davesmith00000 commented on sandbox/src/main/scala/example/Sandbox.scala:
```diff
@@ -209,6 +215,25 @@ object Sandbox extends TyrianApp[Msg, Model]:
+          val handY = 50 + 40 * math.sin(angle)
+          svg(attributes("viewBox" -> "0, 0, 100, 100", "width" -> "300px"))(
+            circle(
+              attributes("cx" -> "50", "cy" -> "50", "r" -> "45", "fill" -> "#0B79CE")
```
The joy of eating your own dog food is that you get to see what sort of meal you made. 😄 
The `circle` tag looks perfect, but it looks like we'll need to generate the SVG attributes like `cx` too?

JPonte replied:
Right, I thought about that. But I was a bit unsure how. I could add to the `HtmlAttributes` but that would mean you could apply them to any other tag, even if not circle or even svg? 
Would it make sense to also split the HtmlAttributes into SvgAttributes and only include those in the SVG object? It doesn't help with enforcing the types but could make it cleaner.

davesmith00000 replied:
I'm afraid that is probably the right thing to do yes. That way when you do the import you get the additional svg attributes.

davesmith00000 commented on project/TagGen.scala:
```diff
@@ -219,6 +233,71 @@ object TagGen {
+
+    def svgTagList  = TagList(svgTags, "SVG")
+
+    def allTags: List[TagList] = List(svgTagList)
```
This seems unnecessary? Only contains the `svgTagList` and is used once?

JPonte replied:
The idea was to have more lists eventually... I don't know what namespaces other than SVG there are but I was expecting some more :D

davesmith00000 replied:
If you get to the end of this journey without adding more, probably best to keep it clean and remove it.

davesmith00000 replied:
Sorry, I just mean the allTags bit. I'm ok with the TagList I think?

davesmith00000 reviewed (commented):
Looking good, well done on the Sandbox modifications. Left a couple of comments.

davesmith00000 commented:
I think that's fine, a best endeavour to catch the most comment attributes is good enough most likely. I never expected that I'd hoovered up every HTML tag either. Having a good go at catching all the main ones and waiting for people to request the other missing ones is a perfectly valid strategy.

davesmith00000 reviewed (approved):
Sorry about the delay, I got side tracked onto Indigo. Approving so I can get this into the next release - many thanks! 😄
